warplib/string_hash.cairo

```diff
@@ -1,33 +1,40 @@
-from starkware.cairo.common.alloc import alloc
-from starkware.cairo.common.cairo_builtins import HashBuiltin
-from starkware.cairo.common.dict_access import DictAccess
-from starkware.cairo.common.hash_chain import hash_chain
-from starkware.cairo.common.memcpy import memcpy
-from warplib.maths.utils import narrow_safe
-from warplib.memory import wm_to_felt_array
-
-func string_hash{pedersen_ptr: HashBuiltin*}(len: felt, ptr: felt*) -> (hashedValue: felt) {
-    alloc_locals;
-    if (len == 0) {
-        return (0,);
-    }
+// from starkware.cairo.common.alloc import alloc
+// from starkware.cairo.common.cairo_builtins import HashBuiltin
+// from starkware.cairo.common.dict_access import DictAccess
+// from starkware.cairo.common.memcpy import memcpy
+// from warplib.maths.utils import narrow_safe
+// from warplib.memory import wm_to_felt_array
+
+
+use array::Array;
+use array::ArrayTrait;
+use array::array_len;
+use integer::u32_to_felt;
+use integer::u32_try_from_felt;
+use debug::print_felt;
+use option::OptionTrait;
 
-    let (temp) = alloc();
-    assert [temp] = len;
-    memcpy(temp + 1, ptr, len);
 
-    let (hashedValue) = hash_chain{hash_ptr=pedersen_ptr}(temp);
-    return (hashedValue,);
+fn hash_chain(arr: @Array::<felt>, pos: u32) -> felt {
+    let next = pos + u32_try_from_felt(1).unwrap();
+    if (next == arr.len()) {
+        return *arr.at(pos);
+    }
+    return pedersen(*arr.at(pos), hash_chain(arr, next));
 }
 
-func wm_string_hash{pedersen_ptr: HashBuiltin*, range_check_ptr, warp_memory: DictAccess*}(
-    mem_loc: felt
-) -> (hashedValue: felt) {
-    alloc_locals;
-    let (len, ptr) = wm_to_felt_array(mem_loc);
 
-    let (hashedValue) = string_hash{pedersen_ptr=pedersen_ptr}(len, ptr);
-    return (hashedValue,);
+fn string_hash(arr: @Array::<felt>) -> felt {
+    return hash_chain(arr, u32_try_from_felt(1).unwrap());
 }
 
+// func wm_string_hash{pedersen_ptr: HashBuiltin*, range_check_ptr, warp_memory: DictAccess*}(
+//     mem_loc: felt
+// ) -> felt {
+//     let (len, ptr) = wm_to_felt_array(mem_loc);
+
+//     let hashedValue = string_hash(len, ptr);
+//     return hashedValue;
+// }
```

Comment:
Normal `string_hash` is used when the array is a normal cairo array, but `wm_string_hash` is used when it is scattered  in a dict access where it transforms it into a normal cairo array, and then applies `string_hash`. I wouldn't mind for it to stay commented a little longer in the code

Reply:
I see, then let's leave it for now. We'll see what to do with that later